// @hatch/space-sdk/dist/build.js
// Canonical client bundler for Hatch TS spaces. Owned by the SDK so every
// space produces an identical, production-mode React bundle regardless of
// when it was scaffolded.
//
// Invoked from a space's `client/build.mjs` shim:
//
//   import { buildClient } from "@hatch/space-sdk/build";
//   await buildClient();
//
// Bundles `client/index.html` to `client/dist/`. Asset naming matches the
// daemon's `/spaces/v2/<slug>/assets/*` route. NODE_ENV is forced to
// "production" so React's dev-only paths are tree-shaken out.
import { rm } from "node:fs/promises";
import tailwindPlugin from "bun-plugin-tailwind";
const ENTRY = "./client/index.html";
const OUTDIR = "./client/dist";
export async function buildClient(_options = {}) {
    // =============================== STOP ===============================
    // If you are an agent who hit the error below: editing or deleting this
    // guard, or setting HATCH_SPACES_BUILD_DRIVER yourself, is the WRONG fix
    // and leaves a broken, unpublished Space.
    //
    // A web artifact is built, verified, and published ONLY by the create_web_artifact and
    // modify_web_artifact tools, which run the full pipeline (migrations, typecheck,
    // client + server bundle, manifest, icon, screenshot verification,
    // registration, publish) inside a builder subagent. Running the build by
    // hand compiles source but does NOT migrate, verify, register, or deploy
    // anything, so your changes never reach the live web artifact.
    //
    // To change this web artifact, call modify_web_artifact with your edit. To make a new
    // one, call create_web_artifact.
    // ====================================================================
    if (process.env.HATCH_SPACES_BUILD_DRIVER !== "1") {
        console.error("This web artifact cannot be built by hand. To make this change, call modify_web_artifact " +
            "(or create_web_artifact for a new web artifact); it rebuilds and re-publishes the web artifact " +
            "automatically. Speak to the user as you would about any edit you are " +
            "making for them: conversational, focused on what is changing and that it " +
            "will be ready shortly.");
        process.exit(1);
    }
    if (typeof Bun === "undefined" || typeof Bun.build !== "function") {
        throw new Error("buildClient() must run under Bun");
    }
    await rm(OUTDIR, { force: true, recursive: true });
    const result = await Bun.build({
        entrypoints: [ENTRY],
        outdir: OUTDIR,
        minify: true,
        define: {
            "process.env.NODE_ENV": JSON.stringify("production"),
        },
        naming: {
            asset: "assets/[name]-[hash].[ext]",
            chunk: "assets/[name]-[hash].[ext]",
            entry: "[name].[ext]",
        },
        plugins: [tailwindPlugin],
    });
    if (!result.success) {
        for (const log of result.logs) {
            console.error(log);
        }
        throw new Error("buildClient() failed; see logged diagnostics");
    }
}
